File: p2a/src/jungle/Game.java
package jungle;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import jungle.pieces.Piece;
import jungle.pieces.Rat;
import jungle.pieces.Tiger;
import jungle.pieces.Lion;
import jungle.squares.Square;
import jungle.squares.Den;
import jungle.squares.PlainSquare;
import jungle.squares.Trap;
import jungle.squares.WaterSquare;

/**
 * 表示丛林游戏的主要游戏逻辑和状态。
 * 管理游戏棋盘、玩家、棋子、回合和游戏规则。
 */
public class Game {
    /** 棋盘高度（行数）。 */
    public static final int HEIGHT = 9;

    /** 棋盘宽度（列数）。 */
    public static final int WIDTH = 7;

    /** 有水方格的行。 */
    public static final int[] WATER_ROWS = {3, 4, 5};

    /** 有水方格的列。 */
    public static final int[] WATER_COLS = {1, 2, 4, 5};

    /** 兽穴的列索引。 */
    public static final int DEN_COL = 3;


    private int currentTurn;
    private Player p0, p1;
    private Player[] players = new Player[2];
    private Square[][] squares = new Square[HEIGHT][WIDTH]; // 游戏棋盘
    private HashMap<Square, Piece> squareToPiece
        = new HashMap<Square, Piece>();

    /**
     * 使用两个玩家构造一个游戏实例。
     * 用棋盘方格初始化游戏棋盘，并将回合设置为第一个玩家。
     * 
     * @param p0 第一个玩家
     * @param p1 第二个玩家
     */
    public Game(Player p0, Player p1) {
        this.p0 = p0;
        this.p1 = p1;
        players[0] = p0;
        players[1] = p1;

        this.initializeBoardSquares();
        this.currentTurn = 0;
    }

    /**
     * 通过在相应位置设置每种方格类型（普通、水、兽穴、陷阱）来初始化游戏棋盘。
     * 棋盘是一个9x7的网格，其中某些方格被指定为水方格，
     * 每个玩家的兽穴和陷阱方格被放置在特定位置。
     */
    private void initializeBoardSquares() {
        // 初始时将所有方格设置为普通方格
        for (int row = 0; row < HEIGHT; row++) {
            for (int col = 0; col < WIDTH; col++) {
                squares[row][col] = new PlainSquare();
            }
        }

        // 基于预定义的行和列设置水方格
        for (int row : WATER_ROWS) {
            for (int col : WATER_COLS) {
                squares[row][col] = new WaterSquare();
            }
        }

        // 为每个玩家设置兽穴和陷阱
        // 设置p0的兽穴和陷阱
        int p0DenRow = 0;
        squares[p0DenRow][DEN_COL] = new Den(p0);
        squares[p0DenRow][DEN_COL - 1] = new Trap(p0);
        squares[p0DenRow][DEN_COL + 1] = new Trap(p0);
        squares[p0DenRow + 1][DEN_COL] = new Trap(p0);

        // 设置p1的兽穴和陷阱
        int p1DenRow = 8;
        squares[p1DenRow][DEN_COL] = new Den(p1);
        squares[p1DenRow][DEN_COL - 1] = new Trap(p1);
        squares[p1DenRow][DEN_COL + 1] = new Trap(p1);
        squares[p1DenRow - 1][DEN_COL] = new Trap(p1);
    }

    /**
     * 根据初始位置和等级为两个玩家添加起始棋子。
     */
    public void addStartingPieces() {
        int[][] initialPieceConfigs = {
                // {行, 列, 等级, 玩家编号}
                // 玩家0
                {2, 0, 1, 0},
                {1, 5, 2, 0},
                {1, 1, 3, 0},
                {2, 4, 4, 0},
                {2, 2, 5, 0},
                {0, 6, 6, 0},
                {0, 0, 7, 0},
                {2, 6, 8, 0},
                // 玩家1
                {6, 6, 1, 1},
                {7, 1, 2, 1},
                {7, 5, 3, 1},
                {6, 2, 4, 1},
                {6, 4, 5, 1},
                {8, 0, 6, 1},
                {8, 6, 7, 1},
                {6, 0, 8, 1},
        };

        for (int[] pieceConig : initialPieceConfigs) {
            int row = pieceConig[0];
            int col = pieceConig[1];
            int rank = pieceConig[2];
            int playerNumber = pieceConig[3];

            addPiece(row, col, rank, playerNumber);
        }
    }

    /**
     * 在棋盘上的给定位置添加一个棋子。
     * 
     * @param row          游戏棋盘上方格的行坐标。
     * @param col          游戏棋盘上方格的列坐标。
     * @param rank         棋子的等级
     * @param playerNumber 拥有该棋子的玩家编号
     */
    public void addPiece(int row, int col, int rank, int playerNumber) {
        Player player = getPlayer(playerNumber);
        Square square = getSquare(row, col);

        Piece piece;
        if (rank == 1) {
            piece = new Rat(player, square);
        } else if (rank == 6) {
            piece = new Tiger(player, square);
        } else if (rank == 7) {
            piece = new Lion(player, square);
        } else {
            piece = new Piece(player, square, rank);
        }

        // 放入squareToPiece映射中
        squareToPiece.put(getSquare(row, col), piece);
    }

    /**
     * 获取棋盘上指定位置的棋子。
     * 
     * @param row 游戏棋盘上的行坐标。
     * @param col 游戏棋盘上的列坐标。
     * @return 给定坐标处的棋子，如果不存在则返回null
     */
    public Piece getPiece(int row, int col) {
        Square square = getSquare(row, col);
        return squareToPiece.get(square);
    }

    /**
     * 将棋子从一个位置移动到另一个位置。
     * 
     * @param fromRow 棋子当前位置的行坐标
     * @param fromCol 棋子当前位置的列坐标
     * @param toRow   目标位置的行坐标
     * @param toCol   目标位置的列坐标
     * @throws IllegalMoveException 如果目标坐标对该棋子不合法
     */
    public void move(int fromRow, int fromCol, int toRow, int toCol) {
        // 检查目标位置是否有效
        if (
            !getLegalMoves(fromRow, fromCol)
                .contains(new Coordinate(toRow, toCol))
        ) {
            throw new IllegalMoveException(
                String.format("(%d, %d) is not legal move", toRow, toCol)
            );
        }

        Piece piece = getPiece(fromRow, fromCol);
        Piece targetPiece = getPiece(toRow, toCol);
        Square sourceSquare = getSquare(fromRow, fromCol);
        Square targetSquare = getSquare(toRow, toCol);

        if (targetPiece != null) {
            targetPiece.beCaptured();
        }

        // 移动棋子
        piece.move(getSquare(toRow, toCol));

        // 更新方格到棋子的映射
        squareToPiece.remove(sourceSquare, piece);
        squareToPiece.put(targetSquare, piece);
        // 轮到另一个玩家
        nextTurn();
    }

    /**
     * 根据玩家编号获取玩家。
     * 
     * @param playerNumber 玩家编号（0或1）
     * @return 具有指定玩家编号的玩家
     * @throws IllegalArgumentException 如果玩家编号无效
     */
    public Player getPlayer(int playerNumber) {
        // 检查玩家编号有效性，有效编号能成功创建玩家
        try {
            Player player = new Player("checkPlayerNumber", playerNumber);
        } catch (IllegalArgumentException e) {
            throw e;
        }

        return playerNumber == p0.getPlayerNumber() ? p0 : p1;
    }

    /**
     * 确定游戏的获胜者。
     * 如果玩家占领了对手的兽穴或者对手没有剩余棋子，
     * 该玩家就获胜。
     * 
     * @return 获胜的玩家，如果还没有获胜者则返回null
     */
    public Player getWinner() {
        if (
            p0.hasCapturedDen() // p0占领了p1的兽穴
            || !p1.hasPieces()  // 或者p1没有剩余棋子
        ) {
            return p0;          // p0获胜
        }

        if (
            p1.hasCapturedDen() // p1占领了p0的兽穴
            || !p0.hasPieces()  // 或者p0没有剩余棋子
        ) {
            return p1;          // p1获胜
        }

        return null;
    }

    /**
     * 检查游戏是否结束。
     * 
     * @return 如果有获胜者则返回true，否则返回false
     */
    public boolean isGameOver() {
        return getWinner() != null;
    }

    /**
     * 获取游戏棋盘上指定坐标的方格。
     * 
     * @param row 棋盘上的行坐标。
     * @param col 棋盘上的列坐标。
     * @return 指定坐标处的方格
     * @throws IndexOutOfBoundsException 如果坐标超出游戏棋盘范围
     */
    public Square getSquare(int row, int col) {
        if (
            (row < 0 || row > HEIGHT - 1)   // 超出高度
            || (col < 0 || col > WIDTH - 1) // 超出宽度
        ) {
            throw new IndexOutOfBoundsException(
                String.format("Coordinate exceed board bounds: (%d, %d)", row, col)
            );
        }

        return squares[row][col];
    }

    /**
     * 获取指定坐标处棋子的合法移动。
     * 
     * @param row 当前行坐标
     * @param col 当前列坐标
     * @return 棋子可以移动的合法坐标列表
     */
    public List<Coordinate> getLegalMoves(int row, int col) {
        List<Coordinate> coordinates = new ArrayList<Coordinate>();

        Player thisTurnPlayer = getPlayer(getTrun() % 2);
        Piece selectedPiece = getPiece(row, col);

        if (
            selectedPiece == null   // 此坐标处没有棋子
            // 或者棋子不属于本回合可移动的玩家
            || !selectedPiece.isOwnedBy(thisTurnPlayer)
            || isGameOver()         // 或者游戏已结束
        ) {
            return coordinates;       // 返回空列表
        }

        // 4个潜在的移动方向
        Coordinate[] directions = {
            new Coordinate(-1, 0),  // 向上
            new Coordinate(1, 0),  // 向下
            new Coordinate(0, -1),  // 向左
            new Coordinate(0, 1),  // 向右
        };

        // 探索在4个方向中每个方向可以移动到的方格
        for (Coordinate direction : directions) {
            Coordinate coordinate = exploreDirection(row,
                                                     col,
                                                     direction.row(),
                                                     direction.col());

            // 如果找到有效方格，添加到列表中
            if (coordinate != null) {
                coordinates.add(coordinate);
            }
        }

        // 返回列表中所有有效方格
        return coordinates;
    }

    /**
     * 从起始位置探索指定方向，查找是否有可以移动到的方格。
     * 处理棋子可以直接移动到相邻方格、游过水或跳跃越过水
     * （如果棋子具有该能力）的情况。
     * 
     * @param row           起始行坐标
     * @param col           起始列坐标
     * @param rowDirection  行方向（例如，-1表示向上，1表示向下）
     * @param colDirection  列方向（例如，-1表示向左，1表示向右）
     * @return 如果移动有效则返回目标坐标，如果不允许则返回null
     * @throws IllegalArgumentException 如果方向无效
     */
    private Coordinate exploreDirection(int row,
                                        int col,
                                        int rowDirection,
                                        int colDirection) {

        // 探索（移动）方向只能是水平或垂直
        if (
            !((rowDirection == -1 || rowDirection == 1)  && colDirection == 0)
            && !(rowDirection == 0 && (colDirection == -1 || colDirection == 1))
        ) {
            throw new IllegalArgumentException(
                "Row and Col Direction, one should be -1 or 1,"
                + " the other should be 0"
            );
        }

        try {
            // 当前探索的方格
            Square exploreSqure = getSquare(row + rowDirection,
                                            col + colDirection);

            Piece selectedPiece = getPiece(row, col);
            Piece distinationPiece = getPiece(row + rowDirection,
                                              col + colDirection);

            if (
                // 探索可以直接移动到的相邻方格
                (
                    // 方格不是水或棋子可以游泳
                    !exploreSqure.isWater() || selectedPiece.canSwim()
                ) && (
                    // 有棋子，但可以击败
                    distinationPiece != null
                    && selectedPiece.canDefeat(distinationPiece)
                    // 或者没有棋子
                    || distinationPiece == null
                )
            ) {
                return new Coordinate(row + rowDirection, col + colDirection);
            } else if (
                // 探索可以跳跃到的方格
                (
                    // 相邻方格是水，且没有棋子
                    exploreSqure.isWater() && distinationPiece == null
                ) && (
                    // 选中的棋子可以垂直跳跃
                    // 且当前探索垂直方向
                    selectedPiece.canLeapVertically() && rowDirection != 0
                    // 或者选中的棋子可以水平跳跃
                    // 且当前探索水平方向
                    || selectedPiece.canLeapHorizontally() && colDirection != 0
                )
            ) {
                int step = 2;
                while (true) {
                    exploreSqure = getSquare(row + rowDirection * step,
                                             col + colDirection * step);
                    distinationPiece = getPiece(row + rowDirection * step,
                                                col + colDirection * step);

                    // 停止探索，如果棋子占据了
                    // 中间的水方格则无法跳跃
                    if (
                        exploreSqure.isWater()
                        && distinationPiece != null
                    ) {
                        break;
                    }

                    // 返回方格不是水，没有棋子或可以被击败
                    if (
                        !exploreSqure.isWater()
                        && (
                            distinationPiece == null
                            || selectedPiece.canDefeat(distinationPiece)
                        )
                    ) {
                        return new Coordinate(row + rowDirection * step,
                                              col + colDirection * step);
                    }

                    // 探索下一个方格
                    step++;
                }
            }
        } catch (IndexOutOfBoundsException e) {
            System.err.println("Exploring: " + e.getMessage());
        }
        return null;
    }

    private int getTrun() {
        return this.currentTurn;
    }

    private void nextTurn() {
        this.currentTurn++;
    }
}

File: p2b/src/jungle/game.py
"""
表示丛林游戏的主要游戏逻辑和状态。
管理游戏棋盘、玩家、棋子、回合和游戏规则。
"""

from typing import List, Optional, Dict
from .player import Player
from .coordinate import Coordinate
from .illegal_move_exception import IllegalMoveException
from .pieces.piece import Piece
from .pieces.rat import Rat
from .pieces.tiger import Tiger
from .pieces.lion import Lion
from .squares.square import Square
from .squares.plain_square import PlainSquare
from .squares.water_square import WaterSquare
from .squares.den import Den
from .squares.trap import Trap


class Game:
    """
    丛林（斗兽棋）的主游戏类。
    
    管理游戏棋盘、玩家、棋子和游戏规则。
    """
    
    # 棋盘高度（行数）
    HEIGHT = 9
    
    # 棋盘宽度（列数）
    WIDTH = 7
    
    # 有水方格的行
    WATER_ROWS = [3, 4, 5]
    
    # 有水方格的列
    WATER_COLS = [1, 2, 4, 5]
    
    # 兽穴的列索引
    DEN_COL = 3
    
    def __init__(self, p0: Player, p1: Player):
        """
        使用两个玩家构造一个游戏实例。
        用棋盘方格初始化游戏棋盘，并将回合设置为第一个玩家。
        
        Args:
            p0: 第一个玩家
            p1: 第二个玩家
        """
        self._p0 = p0
        self._p1 = p1
        self._players = [p0, p1]
        
        # 初始化游戏棋盘
        self._squares: List[List[Square]] = [[None] * self.WIDTH 
                                               for _ in range(self.HEIGHT)]
        self._square_to_piece: Dict[Square, Piece] = {}
        
        self._initialize_board_squares()
        self._current_turn = 0
    
    def _initialize_board_squares(self) -> None:
        """
        通过在各自位置设置每种方格类型（平原、水域、兽穴、陷阱）来初始化游戏棋盘。
        棋盘是一个9x7的网格，其中某些方格被指定为水域方格，
        每个玩家的兽穴和陷阱方格都放置在特定位置。
        """
        # 首先将所有方格设置为普通方格
        for row in range(self.HEIGHT):
            for col in range(self.WIDTH):
                self._squares[row][col] = PlainSquare()
        
        # 根据预定义的行和列设置水域方格
        for row in self.WATER_ROWS:
            for col in self.WATER_COLS:
                self._squares[row][col] = WaterSquare()
        
        # 为每个玩家设置兽穴和陷阱
        # 设置p0的兽穴和陷阱
        p0_den_row = 0
        self._squares[p0_den_row][self.DEN_COL] = Den(self._p0)
        self._squares[p0_den_row][self.DEN_COL - 1] = Trap(self._p0)
        self._squares[p0_den_row][self.DEN_COL + 1] = Trap(self._p0)
        self._squares[p0_den_row + 1][self.DEN_COL] = Trap(self._p0)
        
        # 设置p1的兽穴和陷阱
        p1_den_row = 8
        self._squares[p1_den_row][self.DEN_COL] = Den(self._p1)
        self._squares[p1_den_row][self.DEN_COL - 1] = Trap(self._p1)
        self._squares[p1_den_row][self.DEN_COL + 1] = Trap(self._p1)
        self._squares[p1_den_row - 1][self.DEN_COL] = Trap(self._p1)
    
    def add_starting_pieces(self):
        """
        根据初始位置和等级为两个玩家添加起始棋子。
        """
        initial_piece_configs = [
            # [行, 列, 等级, 玩家编号]
            # 玩家0
            [2, 0, 1, 0],
            [1, 5, 2, 0],
            [1, 1, 3, 0],
            [2, 4, 4, 0],
            [2, 2, 5, 0],
            [0, 6, 6, 0],
            [0, 0, 7, 0],
            [2, 6, 8, 0],
            # 玩家1
            [6, 6, 1, 1],
            [7, 1, 2, 1],
            [7, 5, 3, 1],
            [6, 2, 4, 1],
            [6, 4, 5, 1],
            [8, 0, 6, 1],
            [8, 6, 7, 1],
            [6, 0, 8, 1],
        ]
        
        for piece_config in initial_piece_configs:
            row, col, rank, player_number = piece_config
            self.add_piece(row, col, rank, player_number)
    
    def add_piece(self, row: int, col: int, rank: int, player_number: int):
        """
        在棋盘上的给定位置添加一个棋子。
        
        Args:
            row: 游戏棋盘上方格的行坐标
            col: 游戏棋盘上方格的列坐标
            rank: 棋子的等级
            player_number: 拥有该棋子的玩家编号
        """
        player = self.get_player(player_number)
        square = self.get_square(row, col)
        
        if rank == 1:
            piece = Rat(player, square)
        elif rank == 6:
            piece = Tiger(player, square)
        elif rank == 7:
            piece = Lion(player, square)
        else:
            piece = Piece(player, square, rank)
        
        # 放入方格到棋子的映射中
        self._square_to_piece[self.get_square(row, col)] = piece
    
    def get_piece(self, row: int, col: int) -> Optional[Piece]:
        """
        获取棋盘上指定位置的棋子。
        
        Args:
            row: 游戏棋盘上的行坐标
            col: 游戏棋盘上的列坐标
        
        Returns:
            给定位置的棋子，如果没有棋子则返回None
        """
        square = self.get_square(row, col)
        return self._square_to_piece.get(square)
    
    def move(self, from_row: int, from_col: int, to_row: int, to_col: int):
        """
        将棋子从一个位置移动到另一个位置。
        
        Args:
            from_row: 棋子当前位置的行坐标
            from_col: 棋子当前位置的列坐标
            to_row: 目标位置的行坐标
            to_col: 目标位置的列坐标
            
        Raises:
            IllegalMoveException: 如果目标坐标对于该棋子不合法
        """
        # 检查目标位置是否有效
        if Coordinate(to_row, to_col) not in self.get_legal_moves(from_row, from_col):
            raise IllegalMoveException(
                f"({to_row}, {to_col}) is not legal move"
            )
        
        piece = self.get_piece(from_row, from_col)
        target_piece = self.get_piece(to_row, to_col)
        source_square = self.get_square(from_row, from_col)
        target_square = self.get_square(to_row, to_col)
        
        if target_piece is not None:
            target_piece.be_captured()
        
        # 移动棋子
        piece.move(self.get_square(to_row, to_col))
        
        # 更新方格到棋子的映射
        del self._square_to_piece[source_square]
        self._square_to_piece[target_square] = piece
        
        # 轮到另一个玩家
        self._next_turn()
    
    def get_player(self, player_number: int) -> Player:
        """
        根据玩家编号获取玩家。
        
        Args:
            player_number: 玩家编号（0或1）
            
        Returns:
            具有指定玩家编号的玩家
            
        Raises:
            ValueError: 如果玩家编号无效
        """
        # 检查玩家编号是否有效
        try:
            Player("check_player_number", player_number)
        except ValueError as e:
            raise e
        
        return self._p0 if player_number == self._p0.get_player_number() else self._p1
    
    def get_winner(self) -> Optional[Player]:
        """
        确定游戏的获胜者。
        如果玩家占领对手的兽穴或对手没有棋子剩余，则该玩家获胜。
        
        Returns:
            获胜的玩家，如果还没有获胜者则返回None
        """
        if (
            self._p0.has_captured_den()  # p0占领了p1的兽穴
            or not self._p1.has_pieces()  # 或者p1没有剩余棋子
        ):
            return self._p0  # p0获胜
        
        if (
            self._p1.has_captured_den()  # p1占领了p0的兽穴
            or not self._p0.has_pieces()  # 或者p0没有剩余棋子
        ):
            return self._p1  # p1获胜
        
        return None
    
    def is_game_over(self) -> bool:
        """
        检查游戏是否结束。
        
        Returns:
            如果有获胜者则返回True，否则返回False
        """
        return self.get_winner() is not None
    
    def get_square(self, row: int, col: int) -> Square:
        """
        获取游戏棋盘上指定坐标的方格。
        
        Args:
            row: 棋盘上的行坐标
            col: 棋盘上的列坐标
            
        Returns:
            指定坐标的方格
            
        Raises:
            IndexError: 如果坐标在游戏棋盘外
        """
        if (
            (row < 0 or row > self.HEIGHT - 1)  # 超出高度
            or (col < 0 or col > self.WIDTH - 1)  # 超出宽度
        ):
            raise IndexError(
                f"Coordinate exceed board bounds: ({row}, {col})"
            )
        
        return self._squares[row][col]
    
    def get_legal_moves(self, row: int, col: int) -> List[Coordinate]:
        """
        获取指定坐标处棋子的合法移动。
        
        Args:
            row: 当前行坐标
            col: 当前列坐标
            
        Returns:
            棋子可以移动的合法坐标列表
        """
        coordinates = []
        
        this_turn_player = self.get_player(self._get_turn() % 2)
        selected_piece = self.get_piece(row, col)
        
        if (
            selected_piece is None  # 该坐标没有棋子
            # 或者棋子不属于当前回合可以移动的玩家
            or not selected_piece.is_owned_by(this_turn_player)
            or self.is_game_over()  # 或者游戏已经结束
        ):
            return coordinates  # 返回空列表
        
        # 4个潜在的移动方向
        directions = [
            Coordinate(-1, 0),  # 向上
            Coordinate(1, 0),   # 向下
            Coordinate(0, -1),  # 向左
            Coordinate(0, 1),   # 向右
        ]
        
        # 在4个方向中探索可以移动到的方格
        for direction in directions:
            coordinate = self._explore_direction(
                row,
                col,
                direction.row(),
                direction.col()
            )
            
            # 如果找到有效方格，添加到列表中
            if coordinate is not None:
                coordinates.append(coordinate)
        
        # 返回列表中所有有效方格
        return coordinates
    
    def _explore_direction(
        self,
        row: int,
        col: int,
        row_direction: int,
        col_direction: int
    ) -> Optional[Coordinate]:
        """
        从起始位置探索指定方向，查找是否有可以移动到的方格。
        处理棋子可以直接移动到相邻方格、在水中游泳或跳跃过水域的情况
        （如果棋子具有该能力）。
        
        Args:
            row: 起始行坐标
            col: 起始列坐标
            row_direction: 行方向（例如，-1表示向上，1表示向下）
            col_direction: 列方向（例如，-1表示向左，1表示向右）
            
        Returns:
            如果移动有效则返回目标坐标，如果不允许则返回None
            
        Raises:
            ValueError: 如果方向无效
        """
        # 探索（移动）方向只能是水平或垂直的
        if not (
            ((row_direction == -1 or row_direction == 1) and col_direction == 0)
            or (row_direction == 0 and (col_direction == -1 or col_direction == 1))
        ):
            raise ValueError(
                "Row and Col Direction, one should be -1 or 1, "
                "the other should be 0"
            )
        
        try:
            # 当前探索的方格
            explore_square = self.get_square(row + row_direction,
                                             col + col_direction)
            
            selected_piece = self.get_piece(row, col)
            destination_piece = self.get_piece(row + row_direction,
                                               col + col_direction)
            
            if (
                # 探索直接相邻的可移动方格
                (
                    # 方格不是水域或棋子可以游泳
                    not explore_square.is_water() or selected_piece.can_swim()
                ) and (
                    # 有棋子但可以击败
                    (destination_piece is not None
                     and selected_piece.can_defeat(destination_piece))
                    # 或者没有棋子
                    or destination_piece is None
                )
            ):
                return Coordinate(row + row_direction, col + col_direction)
            elif (
                # 探索可以跳跃到的方格
                (
                    # 相邻方格是水域，并且没有棋子
                    explore_square.is_water() and destination_piece is None
                ) and (
                    # 选中的棋子可以垂直跳跃
                    # 并且当前探索垂直方向
                    (selected_piece.can_leap_vertically() and row_direction != 0)
                    # 或者选中的棋子可以水平跳跃
                    # 并且当前探索水平方向
                    or (selected_piece.can_leap_horizontally() and col_direction != 0)
                )
            ):
                step = 2
                while True:
                    explore_square = self.get_square(row + row_direction * step,
                                                     col + col_direction * step)
                    destination_piece = self.get_piece(row + row_direction * step,
                                                       col + col_direction * step)
                    
                    # 停止探索，如果棋子占据了
                    # 中间的水域则无法跳跃
                    if explore_square.is_water() and destination_piece is not None:
                        break
                    
                    # 返回的方格不是水域，没有棋子或可以被击败
                    if (
                        not explore_square.is_water()
                        and (
                            destination_piece is None
                            or selected_piece.can_defeat(destination_piece)
                        )
                    ):
                        return Coordinate(row + row_direction * step,
                                          col + col_direction * step)
                    
                    # 探索下一个方格
                    step += 1
        except IndexError as e:
            import sys
            print(f"Exploring: {e}", file=sys.stderr)
        
        return None
    
    def _get_turn(self) -> int:
        """获取当前回合数。"""
        return self._current_turn
    
    def _next_turn(self):
        """进入下一回合。"""
        self._current_turn += 1
